zk/conn_test.go

```diff
@@ -0,0 +1,57 @@
+package zk
+
+import (
+	"io/ioutil"
+	"testing"
+	"time"
+)
+
+func TestRecurringReAuthHang(t *testing.T) {
+	sessionTimeout := 2 * time.Second
+
+	finish := make(chan struct{})
+	defer close(finish)
+	go func() {
+		select {
+		case <-finish:
+			return
+		case <-time.After(5 * sessionTimeout):
+			panic("expected not hang")
```

Comment:
We are using here `panic` to signal that the test has failed.

What do you think about re-structuring the test a bit. Instead of launching a go subroutine that panic after some time, we push the:
```
+	currentServer := conn.Server()
+	conn.setDebugCloseRecvLoop(true)
+	zkC.StopServer(currentServer)
+	// wait connect to new zookeeper.
+	for conn.Server() == currentServer && conn.State() != StateHasSession {
+		time.Sleep(100 * time.Millisecond)
+	}
```
Block into a goroutine and wait in the test's goroutine for it to finish, or time out. I *think* that we could even get away without a goroutine and just do the wait-and-timeout in the block:
```
+	for conn.Server() == currentServer && conn.State() != StateHasSession {
+		time.Sleep(100 * time.Millisecond)
+	}
```

This could simplify the test.